### PR TITLE
Taskbaritem > Progressbar / Overlay icon / description / asset mgmt

### PR DESCRIPTION
**Pull request by MyDrift-user**
# Progressbar for running processes

STATE: Finished 
Please accept the PR on your website first, then test this PR and If it looks fine merge it.

## Type of Change
- [x] UI/UX improvement

## Description
Adding Progressbar on taskbar to keep user informed on state of running processes like installing, running tweaks, using MicroWin ....
Polished how and where the assets in winutil are used/saved (at the time logo as png format, logo as ico format, checkmark as png)

Follow [this](https://learn.microsoft.com/en-us/dotnet/api/system.windows.shell.taskbariteminfo?view=windowsdesktop-8.0) link for more info on the subject.

#### Ongoing process
image
image

#### Paused process (unused at the time)
image

#### Error during process
image

#### Indeterminate (stuff that is hard to track right like Microwin):
https://github.com/ChrisTitusTech/winutil/assets/121827219/f6720d51-1a8b-426d-8b27-0cc783971560

#### When the progress has finished and ran without issues, it will show a check on the taskbar until you "activate" (focus) winutil or start another process.
Screenshot 2024-07-16 193828

#### When the progress has finished and had an error, it will show a warning on the taskbar until you "activate" (focus) winutil or start another 
Screenshot 2024-07-16 193754

## Testing
- tested everything listed in Features, and everything worked. 
- When testing you may think that the script is generating errors, which is not the case (on my end), usually it is an error that is also there with the main branch, but goes unnoticed. This will help detecting bugs.

## Features
Support for progress bar;
- Install & Upgradeapps
- uninstall apps
- get installed
- run tweaks
- undo tweaks
- install features
- get microwin iso
- run microwin process

You can now modify the overlay icon using the function.

You can set the description, which is currently not in use.

- If only one app/tweak/feature is selected, it will do an Indeterminatable process state like the video stated before, if multiple are checked it will check & update the process.
- Get Installed apps and the entire microwin section are only in Indeterminatable process state due to maintainability, ease of use and more reasons.
- Upgrade all is currently not supported due to it being impossible to track because the way it is currently updating the applications, after a rework I'll add this feature to the upgeade all process

## Additional Information
For this PR we are using following Properties:
- ProgressState
- ProgressValue
- Overlay is already used for the logo, got a rework
- Description is available but not used at the time

## Checklist
- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no errors/warnings/merge conflicts.

### CODE REVIEW

**Review by og-mrk — CHANGES_REQUESTED** [review]
Overall, I do like this change, as it'll make the experience more polished (even if, for example, less than 10% will notice this progress bar).

A suggestion came to my mind while reading your commit patches is.. why not make an interface for it?

A good (yet, very simple) example for an interface is the`Update-Progress` function in the`Compile.ps1` script. At the core of the function, it's only a 1 line function.. but by doing this, the function can be extended and improved (plus add debug info), all of that without changing the code of each caller (reuse code by making a function).

Also, you could inforce specific values for every parameter, which's pretty nice, IMHO.

**Comment by ChrisTitusTech** [comment]
@MyDrift-user love the idea of putting it in the tray as the GUI can be hard to update while the console is doing the tweaks. Updating the icon tray would be more responsive and a brilliant idea.

File: functions/private/Invoke-WinUtilFeatureInstall.ps1 [hunk 1]
```diff
@@ -10,6 +10,8 @@ function Invoke-WinUtilFeatureInstall {
         $CheckBox
     )
 
+    x = 0
```
Comment (og-mrk) [inline functions/private/Invoke-WinUtilFeatureInstall.ps1:13]: Does this produce any errors?

File: functions/private/Set-WinUtilTaskbarItem.ps1 [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        Set-WinUtilTaskbaritem -value 0.5 -state "Normal"
+        Set-WinUtilTaskbaritem -state "Error"
+        Set-WinUtilTaskbaritem -state "Indeterminate"
+        Set-WinUtilTaskbaritem -description "This is a description"
+        Set-WinUtilTaskbaritem -overlay "C:\path\to\icon.png"
+
+    #>
+
+
+function Set-WinUtilTaskbaritem {
```
Comment (og-mrk) [inline functions/private/Set-WinUtilTaskbarItem.ps1:23]:
```suggestion
function Set-WinUtilTaskbaritem {
    <#

    .SYNOPSIS
        Modifies the Taskbaritem of the WPF Form

    .PARAMETER state & value
        Value can be between 0 and 1, 0 being no progress done yet and 1 being fully completed
        State can be 'None' > No progress, 'Indeterminate' > Without value, 'Normal' > when using value, 'Error' > Red (when using value), 'Paused' > Yellow (when using value)

    .PARAMETER overlay
        Overlay icon to display on the taskbar item

    .EXAMPLE
        Set-WinUtilTaskbaritem -value 0.5 -state "Normal"
        Set-WinUtilTaskbaritem -state "Error"
        Set-WinUtilTaskbaritem -state "Indeterminate"
        Set-WinUtilTaskbaritem -description "This is a description"
        Set-WinUtilTaskbaritem -overlay "C:\path\to\icon.png"

    #>
```

File: functions/private/Set-WinUtilTaskbarItem.ps1 [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    <#
+
+    .SYNOPSIS
+        Modifies the Taskbaritem of the WPF Form
+
+    .PARAMETER state & value
+        Value can be between 0 and 1, 0 being no progress done yet and 1 being fully completed
+        State can be 'None' > No progress, 'Indeterminate' > Without value, 'Normal' > when using value, 'Error' > Red (when using value), 'Paused' > Yellow (when using value)
```
Comment (og-mrk) [inline functions/private/Set-WinUtilTaskbarItem.ps1:9]:
```suggestion
    .PARAMETER State
        State can be 'None' > No progress, 'Indeterminate' > Without value, 'Normal' > when using value, 'Error' > Red (when using value), 'Paused' > Yellow (when using value)
    .PARAMETER Value
        Value can be between 0 and 1, 0 being no progress done yet and 1 being fully completed
```

File: functions/private/Set-WinUtilTaskbarItem.ps1 [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    .PARAMETER state
+        State can be 'None' > No progress, 'Indeterminate' > inf. loading gray, 'Normal' > Gray, 'Error' > Red, 'Paused' > Yellow
+        no value needed:
+        - Set-WinUtilTaskbaritem -state "None"
+        - Set-WinUtilTaskbaritem -state "Indeterminate"
+        value needed:
+        - Set-WinUtilTaskbaritem -state "Error"
+        - Set-WinUtilTaskbaritem -state "Normal"
+        - Set-WinUtilTaskbaritem -state "Paused"
+
```
Comment (og-mrk) [inline functions/private/Set-WinUtilTaskbarItem.ps1:21]:
```suggestion

    .PARAMETER description
        A string to display when the user hovers their mouse over taskbar item.

```

File: functions/private/Set-WinUtilTaskbarItem.ps1 [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    .PARAMETER description
+        Description to display on the taskbar item preview
+        Set-WinUtilTaskbaritem -description "This is a description"
+    #>
+    param (
+        [double]$value,
+        $state,
+        $overlay,
+        $description
+    )
```
Comment (og-mrk) [inline functions/private/Set-WinUtilTaskbarItem.ps1:40]:
```suggestion
    param (
        [double]$value,

        [Parameter(Mandatory=$true)]
        [string]$state,

        [string]$overlay,
        [string]$description
    )
```

File: functions/private/Set-WinUtilTaskbarItem.ps1 [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    if ($state) {
+        switch ($state) {
+            'None' { $sync["Form"].taskbarItemInfo.ProgressState = "None" }
+            'Indeterminate' { $sync["Form"].taskbarItemInfo.ProgressState = "Indeterminate" }
+            'Normal' { $sync["Form"].taskbarItemInfo.ProgressState = "Normal" }
+            'Error' { $sync["Form"].taskbarItemInfo.ProgressState = "Error" }
+            'Paused' { $sync["Form"].taskbarItemInfo.ProgressState = "Paused" }
+            default { write-host "Invalid state" }
```
Comment (og-mrk) [inline functions/private/Set-WinUtilTaskbarItem.ps1:53]:
```suggestion
            default { throw "[Set-WinUtilTaskbarItem] Invalid state" }
```

File: functions/private/convertTo-Bitmap.ps1 [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    # Create a streaming image by streaming the base64 string to a bitmap streamsource
+    $bitmap = New-Object System.Windows.Media.Imaging.BitmapImage
+    $bitmap.BeginInit()
+    $bitmap.StreamSource = [System.IO.MemoryStream][System.Convert]::FromBase64String($base64String)
+    $bitmap.EndInit()
+    $bitmap.Freeze()
+
+    # Return the bitmap object
+    return $bitmap
+}
```
Comment (og-mrk) [inline functions/private/convertTo-Bitmap.ps1:31]: Changing the parameter name from `image` to `imageFilePath` would make it more ddescriptive to other devs.

**Review by og-mrk — APPROVED** [review]
Well done @MyDrift-user

File: functions/private/convertTo-Bitmap.ps1 (hunk 7 above)
Comment (MyDrift-user) [reply functions/private/convertTo-Bitmap.ps1:31]: will be changed on the next push, thanks :)

**Review by og-mrk — APPROVED** [review]
Good commits @MyDrift-user

**Comment by MyDrift-user** [comment]
May meeds changes after the for pr, but I qm currently not at home..

**Review by ChrisTitusTech — APPROVED** [review]

**Pull request merged**